### PR TITLE
Assets refactoring

### PR DESCRIPTION
**Pull request by yeryomenkom**

### CODE REVIEW

File: zmessaging/src/test/scala/com/waz/service/assets2/AssetsServiceSpec.scala [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import com.waz.sync.client.AssetClient2
+import com.waz.sync.client.AssetClient2.{FileWithSha, Metadata, Retention}
+import com.waz.threading.CancellableFuture
+import com.waz.utils.{IoUtils, returning}
+import com.waz.{FilesystemUtils, ZIntegrationMockSpec}
+
+import scala.concurrent.Future
+import scala.util.{Failure, Random, Success}
+
+class AssetsServiceSpec extends ZIntegrationMockSpec {
```
Comment (deanrobertcook) [inline zmessaging/src/test/scala/com/waz/service/assets2/AssetsServiceSpec.scala]: `AssetService`, not Asset**s**Service

File: zmessaging/src/main/scala/com/waz/service/assets2/RawAssetsStorage.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.waz.service.assets2
+
+import com.waz.model.AssetId
+import com.waz.service.assets2.Asset.General
+import com.waz.utils.Storage2
+
+trait RawAssetsStorage extends Storage2[AssetId, RawAsset[General]]
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/service/assets2/RawAssetsStorage.scala]: Not used?

File: zmessaging/src/main/scala/com/waz/service/assets2/AssetsStorage.scala [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import com.waz.utils.TrimmingLruCache.Fixed
+import com.waz.utils.wrappers.{DB, DBCursor}
+import com.waz.utils.{CachedStorage2, CirceJSONSupport, DbStorage2, InMemoryStorage2, Storage2, TrimmingLruCache}
+import io.circe.parser.decode
+import io.circe.syntax._
+import io.circe.{Decoder, Encoder}
+
+import scala.concurrent.ExecutionContext
+
+trait AssetsStorage extends Storage2[AssetId, Asset[General]]
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/service/assets2/AssetsStorage.scala]: This is a good chance to come up with some consistent naming - I don't like the plural form here, so let's keep to the pattern `<Singular><Storage/Service>`: so `AssetStorage` would be better (without the "s")

Comment (deanrobertcook) [reply zmessaging/src/main/scala/com/waz/service/assets2/AssetsStorage.scala]: I also think this will need to extend `ReactiveStorage2`

File: zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+      case Right(response) =>
+        val asset = createAsset(AssetId(response.rId.str), response.token, rawAsset)
+        assetsStorage.save(asset).map(_ => asset).toCancellable
+      case Left(err) =>
+        verbose(s"Error while uploading asset: $err")
+        CancellableFuture.failed(NetworkError(err))
+    }
+  }
+
+  private def createAsset(assetId: AssetId, token: Option[AssetToken], rawAsset: RawAsset[General]): Asset[General] =
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala]: make another `apply` method on the `Asset` class for this logic

File: zmessaging/src/test/scala/com/waz/Config.scala [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+import com.waz.znet2.http.Request.UrlCreator
+import com.waz.znet2.{AuthRequestInterceptor, HttpClientOkHttpImpl}
+
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration._
+
+object Config {
+
+  val testEmailCredentials = EmailCredentials(EmailAddress("mykhailo+5@wire.com"), Password("123456789"))
```
Comment (deanrobertcook) [inline zmessaging/src/test/scala/com/waz/Config.scala]: Maybe it's best to have some sort of hidden config file where everyone can set their own credentials

File: zmessaging/src/main/scala/com/waz/cache2/CacheService.scala [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  def formatSize(size: Long): String = {
+    if (size <= 0) return "0"
+    val units       = Array[String]("B", "kB", "MB", "GB", "TB")
+    val digitGroups = (Math.log10(size) / Math.log10(1024)).toInt
+    new DecimalFormat("#,##0.#").format(size / Math.pow(1024, digitGroups)) + " " + units(digitGroups)
+  }
+
+}
+
+class LruCacheServiceImpl(
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/cache2/CacheService.scala]: Perhaps best to call this `LruFileCacheService(Impl)`, since (at least for me) when I read `LruCache` I think of the Android `LruCache` implementation, which is in memory only

File: zmessaging/src/test/scala/com/waz/Config.scala (hunk 5 above)
Comment (yeryomenkom) [reply zmessaging/src/test/scala/com/waz/Config.scala]: Maybe. I would prefer to have some special user for testing. 
Another problem is that we will create a lot of data during the tests, so it would be nice to have possibility to create a user for each test and remove that user from the backend with all data.
We need to check it with the QA and backend guys. Maybe they know better solutions.

File: zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    assetsStorage.find(asset.id)
+      .flatMap { fromStorage =>
+        if (fromStorage.isEmpty)
+          assetsStorage.save(asset).flatMap(_ => loadFromBackend(asset, callback))
+        else if (asset.localSource.isEmpty)
+          loadFromCache(asset, callback).recoverWith { case _ => loadFromBackend(asset, callback) }
+        else
+          loadFromLocalStorage(asset, callback).recoverWith { case _ => loadFromBackend(asset, callback) }
+      }
+      .toCancellable
```
[the remark this answers is not in the document]
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala]: In this case we will end up with a huge method. I prefer to think that cleanup is just a part of "loading process".

**Comment by deanrobertcook** [comment]
build again

**Comment by deanrobertcook** [comment]
build again

**Comment by deanrobertcook** [comment]
build again

**Comment by deanrobertcook** [comment]
build again

**Comment by JPonte** [comment]
build again

**Comment by deanrobertcook** [comment]
build again

**Comment by deanrobertcook** [comment]
build again

File: zmessaging/src/main/scala/com/waz/utils/crypto/AESUtils.scala [hunk 8]
```diff
@@ -33,10 +34,13 @@ import com.waz.utils.{IoUtils, returning}
   */
 object AESUtils {
 
+  //TODO Why not SecureRandom?
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/utils/crypto/AESUtils.scala]: Remove TODO - I believe this was discussed

File: zmessaging/src/main/scala/com/waz/utils/crypto/AESUtils.scala [hunk 9]
```diff
@@ -33,10 +34,13 @@ import com.waz.utils.{IoUtils, returning}
...
   lazy val randomBytes = new RandomBytes()
 
-  def base64(key: Array[Byte]) = Base64.encodeToString(key, Base64.NO_WRAP | Base64.NO_CLOSE)
-  def base64(key: String) = Base64.decode(key, Base64.NO_WRAP | Base64.NO_CLOSE)
+  @TargetApi(26)
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/utils/crypto/AESUtils.scala]: What are these doing here?

File: zmessaging/src/main/scala/com/waz/utils/crypto/AESUtils.scala [hunk 10]
```diff
@@ -66,33 +70,35 @@ object AESUtils {
   }
 
   def outputStream(key: AESKey, os: OutputStream) = {
-    val iv = randomBytes(16)
+//    val iv = randomBytes(16)
+    val iv = Array.ofDim[Byte](16)
+    ZSecureRandom.nextBytes(iv)
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/utils/crypto/AESUtils.scala]: Undo this change, add another method for the new `CacheService2` if necessary, or we need to figure out how to get this to work on the build servier/JVM

File: zmessaging/src/main/scala/com/waz/utils/CachedStorageImpl.scala [hunk 11]
```diff
@@ -157,7 +310,7 @@ class CachedStorageImpl[K, V](cache: LruCache[K, Option[V]], db: Database)(impli
       }
     }
   } map { case (results, uncached) =>
-    // cache might have changed already at this point, but that would mean the write would have been issued after this read anyway, so we can safely return the outdated values here
+    // cache might have changed already at utils point, but that would mean the write would have been issued after utils read anyway, so we can safely return the outdated values here
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/utils/CachedStorageImpl.scala]: undo

Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/utils/CachedStorageImpl.scala]: done

File: zmessaging/src/main/scala/com/waz/utils/crypto/AESUtils.scala (hunk 10 above)
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/utils/crypto/AESUtils.scala]: done

File: zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala]: We discussed it and decided to keep storing removed assets in database.

File: zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+                      (implicit ec: ExecutionContext) extends AssetService {
+
+  protected def cacheKey(asset: Asset[General]): String = asset.id.str
+
+  private def loadFromBackend(asset: Asset[General], callback: Option[ProgressCallback]): CancellableFuture[InputStream] = {
+    verbose(s"Load asset content from backend. $asset")
+    assetClient.loadAssetContent(asset, callback).flatMap {
+      case Left(err) if err.code == ResponseCode.NotFound =>
+        cache
+          .remove(cacheKey(asset))
```
[the remark this answers is not in the document]
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala]: We discussed it and decided to keep storing removed assets in database.

File: zmessaging/src/main/scala/com/waz/service/assets2/AssetsStorage.scala (hunk 3 above)
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/service/assets2/AssetsStorage.scala]: done

File: zmessaging/src/main/scala/com/waz/service/assets2/RawAssetsStorage.scala (hunk 2 above)
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/service/assets2/RawAssetsStorage.scala]: I need this trait in UI. And in SE in the future.

File: zmessaging/src/test/scala/com/waz/service/assets2/AssetsServiceSpec.scala (hunk 1 above)
Comment (yeryomenkom) [reply zmessaging/src/test/scala/com/waz/service/assets2/AssetsServiceSpec.scala]: done

File: zmessaging/src/main/scala/com/waz/sync/client/AssetClient2.scala [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+  }
+}
+
+object AssetClient2 {
+
+  case class FileWithSha(file: File, sha256: Sha256)
+
+  case class AssetContent(mime: Mime, data: () => InputStream, dataLength: Option[Long])
+
+  case class UploadResponse2(key: RAssetId, expires: Option[Instant], token: Option[AssetToken])
```
[the remark this answers is not in the document]
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/sync/client/AssetClient2.scala]: Done.

File: zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    verbose(s"Load asset content from cache. $asset")
+    cache.get(cacheKey(asset))(asset.encryption)
+      .recoverWith { case err =>
+        verbose(s"Can not load asset content from cache. $err")
+        Future.failed(err)
+      }
+      .toCancellable
+  }
+
+  private def loadFromLocalStorage(asset: Asset[General], callback: Option[ProgressCallback]): CancellableFuture[InputStream] = {
```
[the remark this answers is not in the document]
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/service/assets2/AssetService.scala]: done

File: zmessaging/src/main/scala/com/waz/model/AESKey.scala [hunk 15]
```diff
@@ -29,6 +33,13 @@ object AESKey extends (String => AESKey) {
   val Empty = AESKey("")
 
   def apply(): AESKey = AESUtils.randomKey()
+  //TODO Why not this variant for random key generation?
+  def random: AESKey = {
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/model/AESKey.scala]: Not used - again, randomBytes (libsodium) vs native java libraries

File: zmessaging/src/main/scala/com/waz/model/Dim2.scala [hunk 16]
```diff
@@ -20,7 +20,10 @@ package com.waz.model
 import com.waz.utils.{JsonDecoder, JsonEncoder}
 import org.json.JSONObject
 
-case class Dim2(width: Int, height: Int)
+case class Dim2(width: Int, height: Int) {
+  def swap: Dim2 = Dim2(width = height, height = width)
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/model/Dim2.scala]: Not used

File: zmessaging/src/main/scala/com/waz/model/GenericContent.scala [hunk 17]
```diff
@@ -643,4 +643,10 @@ object GenericContent {
     }
   }
 
+  sealed trait EncryptionAlgorithm2
+  object EncryptionAlgorithm2 {
+    case object AES_CBC extends EncryptionAlgorithm2
+    case object AES_GCM extends EncryptionAlgorithm2
+  }
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/model/GenericContent.scala]: None of this is used

File: zmessaging/src/main/scala/com/waz/model/Uid.scala [hunk 18]
```diff
@@ -117,6 +117,7 @@ case class AssetId(str: String) {
 object AssetId extends (String => AssetId) {
   def apply(): AssetId = Id.random()
 
+  //TODO Why implicit?
```
Comment (deanrobertcook) [inline zmessaging/src/main/scala/com/waz/model/Uid.scala]: It's implicit for the json decoders, I believe. Please remove

File: zmessaging/src/main/scala/com/waz/model/Dim2.scala (hunk 16 above)
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/model/Dim2.scala]: I need it in UI.

File: zmessaging/src/main/scala/com/waz/model/Uid.scala (hunk 18 above)
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/model/Uid.scala]: ok. why object extends function?

File: zmessaging/src/main/scala/com/waz/model/GenericContent.scala (hunk 17 above)
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/model/GenericContent.scala]: removed

File: zmessaging/src/main/scala/com/waz/model/AESKey.scala (hunk 15 above)
Comment (yeryomenkom) [reply zmessaging/src/main/scala/com/waz/model/AESKey.scala]: removed
